app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -1,83 +1,311 @@
+            addBottomDots(position);
+
+            // Change the next button text from next to finish on last slide.
+            btnNext.setText(getString((position == layouts.length - 1) ? R.string.finish : R.string.cont));
+        }
+
+        @Override
+        public void onPageScrolled(int arg0, float arg1, int arg2) {
+
+        }
+
+        @Override
+        public void onPageScrollStateChanged(int arg0) {
+
+        }
+    };
+
+    /**
+     * Making notification bar transparent
+     */
+    private void changeStatusBarColor() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            Window window = getWindow();
+            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
+            window.setStatusBarColor(Color.TRANSPARENT);
+        }
+    }
+
+    /**
+     * View pager adapter
+     */
+    public class ViewPagerAdapter extends PagerAdapter {
+        private LayoutInflater layoutInflater;
+
+        public ViewPagerAdapter() {
+        }
+
+        @Override
+        public Object instantiateItem(ViewGroup container, int position) {
+            layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+            View view = layoutInflater.inflate(layouts[position], container, false);
+
+            /* Slide: storage permission */
+            Button btnGrantStoragePerm = (Button) view.findViewById(R.id.btnGrantStoragePerm);
+            if (btnGrantStoragePerm != null) {
+                btnGrantStoragePerm.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        requestStoragePermission();
+                    }
+                });
+            }
+
+            /* Slide: location permission */
+            Button btnGrantLocationPerm = (Button) view.findViewById(R.id.btnGrantLocationPerm);
+            if (btnGrantLocationPerm != null) {
+                btnGrantLocationPerm.setOnClickListener(new View.OnClickListener() {
+                    @Override
+                    public void onClick(View v) {
+                        requestLocationPermission();
+                    }
+                });
+            }
+
+            container.addView(view);
+            return view;
+        }
+
+        @Override
+        public int getCount() {
+            return layouts.length;
+        }
+
+        @Override
+        public boolean isViewFromObject(View view, Object obj) {
+            return view == obj;
+        }
+
+
+        @Override
+        public void destroyItem(ViewGroup container, int position, Object object) {
+            View view = (View) object;
+            container.removeView(view);
+        }
+    }
+
+    /**
+     * Preconditions:
+     * Storage permission has been granted.
+     */
+    private void startApp() {
+        Boolean doInitialKeyGeneration = !Constants.getConfigFile(this).exists();
```

Comment:
Sidenote: I've replaced the "isFirstStart" check (previously done via preference check) because it's not very reliable detecting if syncthing needs to generate new keys and config. We now use the same indicator we already have in other classes (e.g. ConfigXml). If no config is present (or it got corrupted in a way crucial files are missing) we'll trigger the "key and config generation" process on app start. This gracefully auto-fixes an unrecoverable config break so we don't have to advise users to clear app's data and cache.